### PR TITLE
[REVIEW] Upgrade `arrow` & `pyarrow` to `6.0.1`

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #9645 
This PR upgrades `arrow` & `pyarrow` to `6.0.1` from `5.0.0`.

### CODE REVIEW

**Review by nvdbaranec — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by beckernick** [comment]
Is there any risk to us using the constraint >= arrow/pyarrow 5 rather than requiring 6? This would likely make it much easier to interact with tools that are slower to upgrade.

**Comment by kkraus14** [comment]
> Is there any risk to us using the constraint >= arrow/pyarrow 5 rather than requiring 6? This would likely make it much easier to interact with tools that are slower to upgrade.

We can't do that. We link against the `libarrow` shared library in `libcudf`, and `libarrow` 5.0.0 for example sets its `SONAME` as `libarrow.so.500`. This adds a `NEEDED` entry into `libcudf` of `libarrow.so.500`. If `libarrow` 6.0.0 is installed, the linker will fail to find `libarrow.so.500` and libcudf will give a missing symbols crash.

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by nirandaperera** [comment]
Arrow has new 6.0.1 bug fix release. Could you please use that instead of 6.0.0?

**Comment by nirandaperera** [comment]
> > Is there any risk to us using the constraint >= arrow/pyarrow 5 rather than requiring 6? This would likely make it much easier to interact with tools that are slower to upgrade.
> 
> We can't do that. We link against the `libarrow` shared library in `libcudf`, and `libarrow` 5.0.0 for example sets its `SONAME` as `libarrow.so.500`. This adds a `NEEDED` entry into `libcudf` of `libarrow.so.500`. If `libarrow` 6.0.0 is installed, the linker will fail to find `libarrow.so.500` and libcudf will give a missing symbols crash.

@kkraus14 but looking at this PR, it seems like cudf is not using any new features of arrow v6. IINM cudf only follows the arrow format (which is stable), so can't cudf bind to `libarrow.so` instead of `libarrow.so.xxx`?

**Comment by robertmaynard** [comment]
> > > Is there any risk to us using the constraint >= arrow/pyarrow 5 rather than requiring 6? This would likely make it much easier to interact with tools that are slower to upgrade.
> > 
> > 
> > We can't do that. We link against the `libarrow` shared library in `libcudf`, and `libarrow` 5.0.0 for example sets its `SONAME` as `libarrow.so.500`. This adds a `NEEDED` entry into `libcudf` of `libarrow.so.500`. If `libarrow` 6.0.0 is installed, the linker will fail to find `libarrow.so.500` and libcudf will give a missing symbols crash.
> 
> @kkraus14 but looking at this PR, it seems like cudf is not using any new features of arrow v6. IINM cudf only follows the arrow format (which is stable), so can't cudf bind to `libarrow.so` instead of `libarrow.so.xxx`?

libcudf uses the C++ arrow api and therefore we are bound by the ABI rules ( can't safely drop in 6 for 5 ). You can see an example in `src/interop/from_arrow.cu`

**Review by ajschmidt8 — COMMENTED** [review]
Seems like this PR is still a work in progress, but just a reminder that we'll need to update the `integration` repo below as well.


https://github.com/rapidsai/integration/blob/f0848c2c340de9e511c72d6edcc66e2b310f1c9e/conda/recipes/versions.yaml#L27-L28

File: ci/cpu/build.sh [hunk 1]
```diff
@@ -42,6 +42,9 @@ gpuci_logger "Activate conda env"
 . /opt/conda/etc/profile.d/conda.sh
 conda activate rapids
 
+# TEMP COMMAND
+conda install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda'
```
Comment (galipremsagar) [inline ci/cpu/build.sh:46]:
```suggestion
mamba install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda'
```

File: ci/cpu/build.sh [hunk 2]
```diff
@@ -42,6 +42,9 @@ gpuci_logger "Activate conda env"
 . /opt/conda/etc/profile.d/conda.sh
 conda activate rapids
 
+# TEMP COMMAND
+mamba install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda'
```
Comment (galipremsagar) [inline ci/cpu/build.sh:46]:
```suggestion
#mamba install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda'
```

**Comment by galipremsagar** [comment]
rerun tests

**Comment by Ethyling** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Review by ajschmidt8 — APPROVED** [review]
Approved. CI is passing with Arrow 6 :+1:. We can merge the `integration` PR once CI passes there and then remove the temporary changes from here.

File: ci/cpu/build.sh [hunk 3]
```diff
@@ -46,6 +46,9 @@ gpuci_logger "Activate conda env"
 . /opt/conda/etc/profile.d/conda.sh
 conda activate rapids
 
+# TEMP COMMAND
+#mamba install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda'
+
```
Comment (galipremsagar) [inline ci/cpu/build.sh:51]:
```suggestion
```

File: ci/cpu/build.sh [hunk 4]
```diff
@@ -76,6 +79,9 @@ else
   CONDA_CHANNEL="-c $WORKSPACE/ci/artifacts/cudf/cpu/.conda-bld/"
 fi
 
+gpuci_conda_retry remove --force "arrow-cpp" "pyarrow" "orc" 'arrow-cpp-proc=*=cuda'
+mamba install -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" "orc" 'arrow-cpp-proc=*=cuda'
+
```
Comment (galipremsagar) [inline ci/cpu/build.sh:84]:
```suggestion
```

File: ci/gpu/build.sh [hunk 5]
```diff
@@ -86,16 +86,16 @@ conda list --show-channel-urls
...
                   "rmm=$MINOR_VERSION.*" \
                   "ucx-py=${UCX_PY_VERSION}"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
 # gpuci_conda_retry remove --force rapids-build-env rapids-notebook-env
 # gpuci_mamba_retry install -y "your-pkg=1.0.0"
-
+#gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env orc pyorc
+gpuci_conda_retry remove --force "arrow-cpp" "pyarrow" "orc" 'arrow-cpp-proc=*=cuda'
+gpuci_mamba_retry install --force -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda' "pyorc" "orc" "python-confluent-kafka=1.7.0"
```
Comment (galipremsagar) [inline ci/gpu/build.sh:98]:
```suggestion
```

File: ci/gpu/java.sh [hunk 6]
```diff
@@ -93,8 +91,9 @@ conda activate rapids
 
 # https://docs.rapids.ai/maintainers/depmgmt/
 # gpuci_conda_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
-
+# gpuci_conda_retry install -y "your-pkg=1.0.0"
+gpuci_conda_retry remove --force "arrow-cpp" "pyarrow" "orc" 'arrow-cpp-proc=*=cuda'
+gpuci_mamba_retry install --force -y "arrow-cpp=6.0.1" "pyarrow=6.0.1" 'arrow-cpp-proc=*=cuda' "pyorc" "orc" "python-confluent-kafka=1.7.0"
```
Comment (galipremsagar) [inline ci/gpu/java.sh:96]:
```suggestion
# gpuci_mamba_retry install -y "your-pkg=1.0.0"

```

**Review by kkraus14 — APPROVED** [review]

**Review by rgsl888prabhu — APPROVED** [review]
Looks good, just need to update year in copyright

**Review by bdice — APPROVED** [review]

**Comment by galipremsagar** [comment]
Thanks for the reviews, everyone! @robertmaynard @rgsl888prabhu @bdice 

Thanks @kkraus14 for the help with runtime gxx & cxx libraries.
Thanks @Ethyling for publishing arm packages and @ajschmidt8 for helping with conda issues in CI.
Thanks @jakirkham for the help in resolving `pyorc` packaging issues.

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by Ethyling** [comment]
rerun tests

**Comment by davidwendt** [comment]
@galipremsagar Not sure you are aware of the build error for `QUANTILES_TEST`

```
13:24:38 FAILED: gtests/QUANTILES_TEST 
13:24:38 : && /usr/local/gcc9/bin/g++ -O3 -DNDEBUG  tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/percentile_approx_test.cu.o tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/quantile_test.cpp.o tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/quantiles_test.cpp.o -o gtests/QUANTILES_TEST -L/usr/local/cuda/targets/x86_64-linux/lib/stubs   -L/usr/local/cuda/targets/x86_64-linux/lib -Wl,-rpath,$PREFIX/lib:/opt/conda/envs/rapids/lib:$SRC_DIR/cpp/build:  libcudftestutil.a  /opt/conda/envs/rapids/lib/libgmock_main.so  /opt/conda/envs/rapids/lib/libgtest_main.so  libcudf.so  $PREFIX/lib/libarrow.so.600.1.0  $PREFIX/lib/libarrow_cuda.so.600.1.0  -ldl  $PREFIX/lib/libcudart.so  /usr/lib64/libcuda.so  /opt/conda/envs/rapids/lib/libgmock.so  /opt/conda/envs/rapids/lib/libgtest.so  -pthread  -lcudadevrt  -lcudart_static  -lrt  -lpthread  -ldl && :
13:24:38 tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/percentile_approx_test.cu.o: In function `arrow_percentile_approx(cudf::column_view const&, int, std::vector<double, std::allocator<double> > const&)':
13:24:38 tmpxft_00002e7a_00000000-6_percentile_approx_test.compute_86.cudafe1.cpp:(.text+0x1284): undefined reference to `arrow::internal::TDigest::Quantile(double)'
13:24:38 tmpxft_00002e7a_00000000-6_percentile_approx_test.compute_86.cudafe1.cpp:(.text+0x1488): undefined reference to `arrow::internal::TDigest::MergeInput()'
13:24:38 collect2: error: ld returned 1 exit status
```

Looks like you will need to update this test code before the build will work?
I apologize if this is already known.

**Comment by galipremsagar** [comment]
> @galipremsagar Not sure you are aware of the build error for `QUANTILES_TEST`
> 
> ```
> 13:24:38 FAILED: gtests/QUANTILES_TEST 
> 13:24:38 : && /usr/local/gcc9/bin/g++ -O3 -DNDEBUG  tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/percentile_approx_test.cu.o tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/quantile_test.cpp.o tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/quantiles_test.cpp.o -o gtests/QUANTILES_TEST -L/usr/local/cuda/targets/x86_64-linux/lib/stubs   -L/usr/local/cuda/targets/x86_64-linux/lib -Wl,-rpath,$PREFIX/lib:/opt/conda/envs/rapids/lib:$SRC_DIR/cpp/build:  libcudftestutil.a  /opt/conda/envs/rapids/lib/libgmock_main.so  /opt/conda/envs/rapids/lib/libgtest_main.so  libcudf.so  $PREFIX/lib/libarrow.so.600.1.0  $PREFIX/lib/libarrow_cuda.so.600.1.0  -ldl  $PREFIX/lib/libcudart.so  /usr/lib64/libcuda.so  /opt/conda/envs/rapids/lib/libgmock.so  /opt/conda/envs/rapids/lib/libgtest.so  -pthread  -lcudadevrt  -lcudart_static  -lrt  -lpthread  -ldl && :
> 13:24:38 tests/CMakeFiles/QUANTILES_TEST.dir/quantiles/percentile_approx_test.cu.o: In function `arrow_percentile_approx(cudf::column_view const&, int, std::vector<double, std::allocator<double> > const&)':
> 13:24:38 tmpxft_00002e7a_00000000-6_percentile_approx_test.compute_86.cudafe1.cpp:(.text+0x1284): undefined reference to `arrow::internal::TDigest::Quantile(double)'
> 13:24:38 tmpxft_00002e7a_00000000-6_percentile_approx_test.compute_86.cudafe1.cpp:(.text+0x1488): undefined reference to `arrow::internal::TDigest::MergeInput()'
> 13:24:38 collect2: error: ld returned 1 exit status
> ```
> 
> Looks like you will need to update this test code before the build will work? I apologize if this is already known.

Have seen this undefined symbol issue, We don't see this error locally. It appears to be happening in CI if we try to upgrade arrow to 6.0 from 5.0 in ci, if we force remove 5.0 arrow and install 6.0 we don't get this problem. So @Ethyling upgraded our build images too with arrow 6.0 and reran the tests.

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**